File: src/rate_of_closure/web/src/model/capabilityOptimizer.ts
/** Deterministic robust optimization over clubs and delivery capabilities. */

import {
  type CapabilityEvaluator,
  type CapabilityObjective,
  type ClubCapability,
  type OptimizationAlternative,
  type OptimizationRequest,
  type OptimizationResult,
  type PlayerCapabilityProfile,
} from "./capabilityContract";
import { covarianceFactor, primeBases, radicalInverse } from "./capabilityMath";
import { contains, type TargetRegionTs } from "./targets";

const FAILURE_PENALTY = 1_000_000;
const BOUNDARY_TOLERANCE = 1e-9;
const PROVENANCE = Object.freeze({
  ensemble: "deterministic-correlated-low-discrepancy/v1",
  flight_metrics: "ball-flight-metrics/v1",
  optimizer: "capability-optimizer/v1",
  target_geometry: "swing_sim.solver.targets/v1",
});

interface Landing { readonly carryM: number; readonly offlineM: number }
interface Counts { readonly completed: number; readonly noImpact: number; readonly failed: number }
interface CandidateSummary { readonly alternative: OptimizationAlternative | null; readonly counts: Counts }
interface RiskMetrics {
  readonly meanCarryM: number; readonly expectedMissM: number;
  readonly holdProbability: number; readonly dispersionRmsM: number;
  readonly cvarMissM: number; readonly downsideCarryM: number;
}

const clubById = (profile: PlayerCapabilityProfile, clubId: string): ClubCapability => {
  const club = profile.clubs.find((item) => item.clubId === clubId);
  if (!club) throw new RangeError(`unknown club_id: ${clubId}`);
  return club;
};

const targetRegion = (request: OptimizationRequest): TargetRegionTs => Object.freeze({
  kind: request.target.kind, distanceM: request.target.distanceM,
  lateralM: request.target.lateralM, radiusM: request.target.radiusM,
  bandHalfLengthM: request.target.bandHalfLengthM, halfWidthM: request.target.halfWidthM,
});

const candidateParameters = (
  club: ClubCapability, candidateIndex: number, seed: number,
): Record<string, number> => {
  if (candidateIndex === 0) return Object.fromEntries(club.parameters.map((item) => [item.parameterId, item.baseline]));
  const sequenceIndex = candidateIndex + seed;
  const bases = primeBases(club.parameters.length);
  return Object.fromEntries(club.parameters.map((item, index) => [
    item.parameterId,
    item.lowerBound + radicalInverse(sequenceIndex, bases[index]) * (item.upperBound - item.lowerBound),
  ]));
};

const dimensionalCovariance = (club: ClubCapability): number[][] => {
  if (club.matrixKind === "covariance") return club.matrix.map((row) => [...row]);
  return club.matrix.map((row, rowIndex) => row.map((entry, columnIndex) =>
    entry * club.parameters[rowIndex].standardDeviation * club.parameters[columnIndex].standardDeviation));
};

const perturbedParameters = (
  club: ClubCapability, nominal: Readonly<Record<string, number>>, sampleIndex: number, seed: number,
): Record<string, number> => {
  const bases = primeBases(club.parameters.length);
  const sequenceIndex = sampleIndex + seed + 1;
  const independent = bases.map((base) => Math.sqrt(3) * (2 * radicalInverse(sequenceIndex, base) - 1));
  const factor = covarianceFactor(dimensionalCovariance(club));
  const correlated = factor.map((row) => row.reduce((sum, value, index) => sum + value * independent[index], 0));
  return Object.fromEntries(club.parameters.map((item, index) => [
    item.parameterId,
    Math.min(item.upperBound, Math.max(item.lowerBound, nominal[item.parameterId] + item.bias + correlated[index])),
  ]));
};

const parseLanding = (evaluation: ReturnType<CapabilityEvaluator>): Landing | null => {
  if (evaluation.status !== "complete") return null;
  const carry = evaluation.metrics.find((item) => item.metricId === "carry_distance")?.value;
  const offline = evaluation.metrics.find((item) => item.metricId === "carry_offline")?.value;
  if (!Number.isFinite(carry) || !Number.isFinite(offline)) return null;
  return Object.freeze({ carryM: carry as number, offlineM: offline as number });
};

const tailMean = (values: readonly number[], alpha: number, reverse: boolean): number => {
  const count = Math.max(1, Math.ceil(values.length * (1 - alpha)));
  const ordered = [...values].sort((left, right) => reverse ? right - left : left - right);
  return ordered.slice(0, count).reduce((sum, value) => sum + value, 0) / count;
};

const limitingConstraints = (
  club: ClubCapability,
  nominal: Readonly<Record<string, number>>,
  successFraction: number,
  extrapolated: boolean,
  request: OptimizationRequest,
): readonly string[] => {
  const limiting = club.parameters.flatMap((item) => {
    if (Math.abs(nominal[item.parameterId] - item.lowerBound) <= BOUNDARY_TOLERANCE) return [`${item.parameterId}:lower_safe_bound`];
    if (Math.abs(nominal[item.parameterId] - item.upperBound) <= BOUNDARY_TOLERANCE) return [`${item.parameterId}:upper_safe_bound`];
    return [];
  });
  if (successFraction < request.minimumSuccessFraction) limiting.push("minimum_success_fraction");
  if (extrapolated) limiting.push("evidence_envelope");
  return Object.freeze(limiting);
};

const objectiveScore = (
  objective: CapabilityObjective,
  risk: RiskMetrics,
  targetDistance: number,
): number => {
  if (objective === "maximize_carry") return -risk.meanCarryM;
  if (objective === "minimize_expected_miss") return risk.expectedMissM;
  if (objective === "maximize_target_hold") return -risk.holdProbability + risk.expectedMissM * 1e-6;
  if (objective === "minimize_variability") return risk.dispersionRmsM;
  if (objective === "minimize_downside") return risk.cvarMissM + risk.downsideCarryM;
  return Math.abs(risk.meanCarryM - targetDistance) + risk.dispersionRmsM;
};

const summarize = (
  club: ClubCapability,
  nominal: Readonly<Record<string, number>>,
  landings: readonly Landing[],
  counts: Counts,
  profile: PlayerCapabilityProfile,
  request: OptimizationRequest,
): OptimizationAlternative | null => {
  if (landings.length === 0) return null;
  const target = targetRegion(request);
  const centerOffline = target.kind === "green" ? target.lateralM : 0;
  const carries = landings.map((item) => item.carryM);
  const offlines = landings.map((item) => item.offlineM);
  const misses = landings.map((item) => Math.hypot(item.carryM - target.distanceM, item.offlineM - centerOffline));
  const meanCarry = carries.reduce((sum, value) => sum + value, 0) / carries.length;
  const meanOffline = offlines.reduce((sum, value) => sum + value, 0) / offlines.length;
  const expectedMiss = misses.reduce((sum, value) => sum + value, 0) / misses.length;
  const dispersion = Math.sqrt(landings.reduce((sum, item) =>
    sum + (item.carryM - meanCarry) ** 2 + (item.offlineM - meanOffline) ** 2, 0) / landings.length);
  const holdProbability = landings.filter((item) => contains(target, item.carryM, item.offlineM)).length / landings.length;
  const cvarMissM = tailMean(misses, request.cvarAlpha, true);
  const downsideCarryM = Math.max(0, target.distanceM - tailMean(carries, request.cvarAlpha, false));
  const risk = { meanCarryM: meanCarry, expectedMissM: expectedMiss, holdProbability, dispersionRmsM: dispersion, cvarMissM, downsideCarryM };
  const successFraction = counts.completed / request.ensembleSize;
  const failureFraction = 1 - successFraction;
  const extrapolated = club.parameters.some((item) =>
    nominal[item.parameterId] < item.evidenceLowerBound || nominal[item.parameterId] > item.evidenceUpperBound);
  const confidence = profile.confidence * club.confidence * successFraction * (extrapolated ? 0.5 : 1);
  let score = objectiveScore(request.objective, risk, target.distanceM);
  if (successFraction < request.minimumSuccessFraction) score += FAILURE_PENALTY * (request.minimumSuccessFraction - successFraction);
  return Object.freeze({
    rank: 1, clubId: club.clubId,
    parameters: Object.freeze(Object.entries(nominal).map(([parameterId, value]) => Object.freeze({ parameterId, value }))),
    score, meanCarryM: meanCarry, expectedMissM: expectedMiss, dispersionRmsM: dispersion,
    targetHoldProbability: holdProbability, cvarMissM, downsideCarryM,
    sampleCount: request.ensembleSize, successfulCount: counts.completed,
    noImpactCount: counts.noImpact, failedCount: counts.failed, failureFraction, confidence,
    limitingConstraints: limitingConstraints(club, nominal, successFraction, extrapolated, request),
    extrapolated, paretoEfficient: false,
  });
};

const evaluateCandidate = (
  club: ClubCapability,
  nominal: Readonly<Record<string, number>>,
  profile: PlayerCapabilityProfile,
  request: OptimizationRequest,
  evaluator: CapabilityEvaluator,
): CandidateSummary => {
  const landings: Landing[] = [];
  let counts: Counts = { completed: 0, noImpact: 0, failed: 0 };
  for (let sampleIndex = 0; sampleIndex < request.ensembleSize; sampleIndex += 1) {
    let evaluation: ReturnType<CapabilityEvaluator>;
    try { evaluation = evaluator(club.clubId, perturbedParameters(club, nominal, sampleIndex, request.seed)); }
    catch { counts = { ...counts, failed: counts.failed + 1 }; continue; }
    const landing = parseLanding(evaluation);
    if (!landing) {
      counts = evaluation.status === "no_impact"
        ? { ...counts, noImpact: counts.noImpact + 1 }
        : { ...counts, failed: counts.failed + 1 };
      continue;
    }
    counts = { ...counts, completed: counts.completed + 1 };
    landings.push(landing);
  }
  return { alternative: summarize(club, nominal, landings, counts, profile, request), counts };
};

const paretoMark = (
  alternatives: readonly OptimizationAlternative[], request: OptimizationRequest,
): readonly OptimizationAlternative[] => {
  if (request.objective !== "distance_control_pareto") return alternatives;
  return alternatives.map((candidate) => {
    const distanceError = Math.abs(candidate.meanCarryM - request.target.distanceM);
    const dominated = alternatives.some((other) => other !== candidate
      && Math.abs(other.meanCarryM - request.target.distanceM) <= distanceError
      && other.dispersionRmsM <= candidate.dispersionRmsM
      && (Math.abs(other.meanCarryM - request.target.distanceM) < distanceError
        || other.dispersionRmsM < candidate.dispersionRmsM));
    return Object.freeze({ ...candidate, paretoEfficient: !dominated });
  });
};

/** Rank robust shot alternatives while delegating all flight physics to `evaluator`. */
export function optimizeCapability(
  profile: PlayerCapabilityProfile,
  request: OptimizationRequest,
  evaluator: CapabilityEvaluator,
): OptimizationResult {
  const clubs = request.clubIds.map((clubId) => clubById(profile, clubId));
  const indices = new Map(clubs.map((club) => [club.clubId, 0]));
  const alternatives: OptimizationAlternative[] = [];
  let aggregate: Counts = { completed: 0, noImpact: 0, failed: 0 };
  for (let evaluationIndex = 0; evaluationIndex < request.candidateBudget; evaluationIndex += 1) {
    const club = clubs[evaluationIndex % clubs.length];
    const candidateIndex = indices.get(club.clubId) as number;
    indices.set(club.clubId, candidateIndex + 1);
    const summary = evaluateCandidate(club, candidateParameters(club, candidateIndex, request.seed), profile, request, evaluator);
    aggregate = { completed: aggregate.completed + summary.counts.completed, noImpact: aggregate.noImpact + summary.counts.noImpact, failed: aggregate.failed + summary.counts.failed };
    if (summary.alternative) alternatives.push(summary.alternative);
  }
  const marked = [...paretoMark(alternatives, request)];
  marked.sort((left, right) => {
    if (request.objective === "distance_control_pareto" && left.paretoEfficient !== right.paretoEfficient) return left.paretoEfficient ? -1 : 1;
    return left.score - right.score || left.clubId.localeCompare(right.clubId)
      || JSON.stringify(left.parameters).localeCompare(JSON.stringify(right.parameters));
  });
  const ranked = Object.freeze(marked.slice(0, request.alternativesCount)
    .map((item, index) => Object.freeze({ ...item, rank: index + 1 })));
  return Object.freeze({
    schemaVersion: "capability-optimization-result/v1", problemId: request.problemId,
    status: ranked.length > 0 ? "solved" : "nonconverged", alternatives: ranked,
    evaluationsAttempted: request.candidateBudget * request.ensembleSize,
    evaluationsCompleted: aggregate.completed, noImpactCount: aggregate.noImpact,
    failedCount: aggregate.failed, provenance: PROVENANCE,
  });
}

File: src/shared/python/swing_sim/flight/capability_optimizer.py
"""Deterministic robust optimization over clubs and delivery capabilities."""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, replace

from ._capability_observation_runtime import (
    _CandidateIdentity,
    _evaluate_sample,
    _OptimizationContext,
    _SampleAttempt,
)
from ._capability_optimizer_metrics import (
    _Landing,
    _landing,
    _risk_metrics,
    _score,
)
from .capability_contract import (
    CapabilityObjective,
    ClubCapability,
    OptimizationAlternative,
    OptimizationRequest,
    OptimizationResult,
    PlayerCapabilityProfile,
)
from .capability_observation import (
    CapabilityOptimizationCancelled,
    CapabilityOptimizationHooks,
    CapabilitySampleStatus,
)
from .capability_sampling import (
    sample_candidate_parameters,
    sample_perturbed_parameters,
)
from .inverse_contract import EvaluationStatus, SolverEvaluation

CapabilityEvaluator = Callable[[str, dict[str, float]], SolverEvaluation]
_FAILURE_PENALTY = 1_000_000.0
_BOUNDARY_TOLERANCE = 1e-9
_PROVENANCE = (
    ("ensemble", "deterministic-correlated-low-discrepancy/v1"),
    ("flight_metrics", "ball-flight-metrics/v1"),
    ("optimizer", "capability-optimizer/v1"),
    ("target_geometry", "swing_sim.solver.targets/v1"),
)


@dataclass(frozen=True)
class _Counts:
    completed: int = 0
    no_impact: int = 0
    failed: int = 0

    def add(self, status: EvaluationStatus) -> _Counts:
        """Return counts with one typed outcome added."""
        if status is EvaluationStatus.COMPLETE:
            return replace(self, completed=self.completed + 1)
        if status is EvaluationStatus.NO_IMPACT:
            return replace(self, no_impact=self.no_impact + 1)
        return replace(self, failed=self.failed + 1)


@dataclass(frozen=True)
class _CandidateSamples:
    club: ClubCapability
    nominal: dict[str, float]
    landings: tuple[_Landing, ...]
    counts: _Counts


def _limiting_constraints(
    samples: _CandidateSamples,
    success_fraction: float,
    extrapolated: bool,
    request: OptimizationRequest,
) -> tuple[str, ...]:
    limiting: list[str] = []
    for item in samples.club.parameters:
        value = samples.nominal[item.parameter_id]
        if abs(value - item.lower_bound) <= _BOUNDARY_TOLERANCE:
            limiting.append(f"{item.parameter_id}:lower_safe_bound")
        elif abs(value - item.upper_bound) <= _BOUNDARY_TOLERANCE:
            limiting.append(f"{item.parameter_id}:upper_safe_bound")
    if success_fraction < request.minimum_success_fraction:
        limiting.append("minimum_success_fraction")
    if extrapolated:
        limiting.append("evidence_envelope")
    return tuple(limiting)


def _summarize(
    samples: _CandidateSamples, context: _OptimizationContext
) -> OptimizationAlternative | None:
    profile, request = context.profile, context.request
    if not samples.landings:
        return None
    risk = _risk_metrics(samples.landings, request)
    success_fraction = samples.counts.completed / request.ensemble_size
    failure_fraction = 1.0 - success_fraction
    extrapolated = any(
        not item.evidence_lower_bound
        <= samples.nominal[item.parameter_id]
        <= item.evidence_upper_bound
        for item in samples.club.parameters
    )
    confidence = (
        profile.confidence
        * samples.club.confidence
        * success_fraction
        * (0.5 if extrapolated else 1.0)
    )
    constraints = _limiting_constraints(
        samples, success_fraction, extrapolated, request
    )
    score = _score(request.objective, risk, request.target.distance_m)
    if success_fraction < request.minimum_success_fraction:
        score += _FAILURE_PENALTY * (
            request.minimum_success_fraction - success_fraction
        )
    return OptimizationAlternative(
        1,
        samples.club.club_id,
        tuple(samples.nominal.items()),
        score,
        risk.mean_carry_m,
        risk.expected_miss_m,
        risk.dispersion_rms_m,
        risk.hold_probability,
        risk.cvar_miss_m,
        risk.downside_carry_m,
        request.ensemble_size,
        samples.counts.completed,
        samples.counts.no_impact,
        samples.counts.failed,
        failure_fraction,
        confidence,
        constraints,
        extrapolated,
        False,
    )


def _evaluate_candidate(
    club: ClubCapability,
    nominal: dict[str, float],
    identity: _CandidateIdentity,
    context: _OptimizationContext,
) -> tuple[OptimizationAlternative | None, _Counts]:
    landings: list[_Landing] = []
    counts = _Counts()
    request = context.request
    for sample_index in range(request.ensemble_size):
        attempt_ordinal = (
            identity.candidate_ordinal * request.ensemble_size + sample_index
        )
        if context.hooks.should_cancel is not None and context.hooks.should_cancel():
            raise CapabilityOptimizationCancelled(attempt_ordinal, context.total_count)
        parameters = sample_perturbed_parameters(
            club, nominal, sample_index, request.seed
        )
        outcome = _evaluate_sample(
            _SampleAttempt(
                request.problem_id,
                attempt_ordinal,
                context.total_count,
                identity.candidate_ordinal,
                identity.club_candidate_ordinal,
                sample_index,
                club,
                nominal,
                parameters,
            ),
            context.evaluator,
            context.hooks,
        )
        evaluation = outcome.evaluation
        status = EvaluationStatus(outcome.effective_status.value)
        counts = counts.add(status)
        if outcome.effective_status is not CapabilitySampleStatus.COMPLETE:
            continue
        if evaluation is None:
            raise RuntimeError("normalized complete sample has no evaluation")
        landing = _landing(evaluation)
        if landing is None:
            raise RuntimeError("normalized complete sample has no finite landing")
        landings.append(landing)
    samples = _CandidateSamples(club, nominal, tuple(landings), counts)
    return _summarize(samples, context), counts


def _pareto_mark(
    alternatives: list[OptimizationAlternative], request: OptimizationRequest
) -> list[OptimizationAlternative]:
    if request.objective is not CapabilityObjective.DISTANCE_CONTROL_PARETO:
        return alternatives
    target = request.target.distance_m
    marked: list[OptimizationAlternative] = []
    for candidate in alternatives:
        distance_error = abs(candidate.mean_carry_m - target)
        dominated = any(
            other is not candidate
            and abs(other.mean_carry_m - target) <= distance_error
            and other.dispersion_rms_m <= candidate.dispersion_rms_m
            and (
                abs(other.mean_carry_m - target) < distance_error
                or other.dispersion_rms_m < candidate.dispersion_rms_m
            )
            for other in alternatives
        )
        marked.append(replace(candidate, pareto_efficient=not dominated))
    return marked


def _collect_alternatives(
    clubs: list[ClubCapability], context: _OptimizationContext
) -> tuple[list[OptimizationAlternative], _Counts]:
    alternatives: list[OptimizationAlternative] = []
    aggregate = _Counts()
    per_club_indices = {club.club_id: 0 for club in clubs}
    for candidate_ordinal in range(context.request.candidate_budget):
        club = clubs[candidate_ordinal % len(clubs)]
        club_ordinal = per_club_indices[club.club_id]
        per_club_indices[club.club_id] += 1
        nominal = sample_candidate_parameters(club, club_ordinal, context.request.seed)
        alternative, counts = _evaluate_candidate(
            club,
            nominal,
            _CandidateIdentity(candidate_ordinal, club_ordinal),
            context,
        )
        aggregate = _Counts(
            aggregate.completed + counts.completed,
            aggregate.no_impact + counts.no_impact,
            aggregate.failed + counts.failed,
        )
        if alternative is not None:
            alternatives.append(alternative)
    return alternatives, aggregate


def optimize_capability(
    profile: PlayerCapabilityProfile,
    request: OptimizationRequest,
    evaluator: CapabilityEvaluator,
    *,
    hooks: CapabilityOptimizationHooks | None = None,
) -> OptimizationResult:
    """Rank robust shot alternatives using injected canonical flight evaluations.

    Preconditions:
        Requested clubs exist in ``profile`` and evaluator outputs use canonical
        carry-distance and carry-offline metrics.
    Postconditions:
        Sampling, rankings, confidence, failure diagnostics, and Pareto flags are
        deterministic for a deterministic evaluator.
    """
    clubs = [profile.club(club_id) for club_id in request.club_ids]
    total_count = request.candidate_budget * request.ensemble_size
    active_hooks = hooks or CapabilityOptimizationHooks()
    context = _OptimizationContext(profile, request, evaluator, active_hooks)
    alternatives, aggregate = _collect_alternatives(clubs, context)
    alternatives = _pareto_mark(alternatives, request)
    alternatives.sort(
        key=lambda item: (
            (
                not item.pareto_efficient
                if request.objective is CapabilityObjective.DISTANCE_CONTROL_PARETO
                else False
            ),
            item.score,
            item.club_id,
            item.parameters,
        )
    )
    ranked = tuple(
        replace(item, rank=index + 1)
        for index, item in enumerate(alternatives[: request.alternatives_count])
    )
    status = "solved" if ranked else "nonconverged"
    return OptimizationResult(
        request.problem_id,
        status,
        ranked,
        total_count,
        aggregate.completed,
        aggregate.no_impact,
        aggregate.failed,
        _PROVENANCE,
    )


__all__ = ["CapabilityEvaluator", "optimize_capability"]
